bridges/snowbridge/pallets/system-frontend/src/lib.rs

```diff
@@ -220,25 +242,116 @@ pub mod pallet {
 			Self::deposit_event(Event::ExportOperatingModeChanged { mode });
 			Ok(())
 		}
+
+		/// Add an additional relayer tip for a committed message identified by `message_id`.
+		/// The tip asset will be swapped for ether.
+		#[pallet::call_index(2)]
+		#[pallet::weight(
+			T::WeightInfo::add_tip()
+				.saturating_add(T::BackendWeightInfo::transact_add_tip())
+		)]
+		pub fn add_tip(
+			origin: OriginFor<T>,
+			message_id: MessageId,
+			asset: Asset,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let ether_location = T::EthereumLocation::get();
+			let (tip_asset_location, tip_amount) = match asset {
+				Asset { id: AssetId(ref loc), fun: Fungibility::Fungible(amount) } => (loc, amount),
+				_ => return Err(Error::<T>::UnsupportedAsset.into()),
+			};
+
+			let ether_gained = if *tip_asset_location != ether_location {
+				Self::swap_and_burn(
+					who.clone(),
+					tip_asset_location.clone(),
+					ether_location,
+					tip_amount,
+				)
+				.map_err(|_| Error::<T>::SwapError)?
+			} else {
+				tip_amount
+			};
+
+			// Send the tip details to BH to be allocated to the reward in the Inbound/Outbound
+			// pallet
+			let dest = T::BridgeHubLocation::get();
+			let call = Self::build_add_tip_call(who.clone(), message_id.clone(), ether_gained);
+			let remote_xcm = Self::build_remote_xcm(&call);
+			let local_pallet_origin: Location = T::PalletLocation::get().into();
+			let who_location = Self::account_to_location(who)?;
+
+			let xcm_message_id =
+				Self::send_xcm(local_pallet_origin, who_location, dest.clone(), remote_xcm.clone())
+					.map_err(|error| Error::<T>::from(error))?;
+
+			Self::deposit_event(Event::<T>::RewardTipAdded {
+				origin: T::PalletLocation::get().into(),
+				message_id,
+				asset,
+				xcm_message_id,
+			});
+
+			Ok(())
+		}
 	}
 
 	impl<T: Config> Pallet<T> {
-		fn send_xcm(origin: Location, dest: Location, xcm: Xcm<()>) -> Result<XcmHash, SendError> {
+		fn send_xcm(
+			origin: Location,
+			sender: Location,
+			dest: Location,
+			xcm: Xcm<()>,
+		) -> Result<XcmHash, SendError> {
 			let is_waived =
 				<T::XcmExecutor as FeeManager>::is_waived(Some(&origin), FeeReason::ChargeFees);
 			let (ticket, price) = validate_send::<T::XcmSender>(dest, xcm.clone())?;
 			if !is_waived {
-				T::XcmExecutor::charge_fees(origin, price).map_err(|_| SendError::Fees)?;
+				T::XcmExecutor::charge_fees(sender, price).map_err(|_| SendError::Fees)?;
 			}
 			T::XcmSender::deliver(ticket)
 		}
 
+		fn swap_and_burn(
```

Comment:
Comments please :)

Reply:
Resolved in https://github.com/vgeddes/polkadot-sdk/commit/a0fda22f8e4a362f46b101909969750fd2cd6ac7